bare_header/device.c++

```diff
@@ -0,0 +1,95 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "bare_header/device.h"
+
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <string>
+
+Device::Device(std::ostream &os, const fdt &dtb, std::string compat_string)
+  : os(os), dtb(dtb), compat_string(compat_string)
+{}
+
+void Device::emit_comment(const node &n) {
+  os << "/* From " << n.name() << " */\n";
+}
+
+string Device::def_handle(const node &n) {
+  string name = compat_string;
+  string instance = n.instance();
+
+  std::transform(name.begin(), name.end(), name.begin(),
+      [](unsigned char c) -> char { 
+	if(c == ',' || c == '-') {
+	  return '_';
+	}
+	return toupper(c);
+      });
+  std::transform(instance.begin(), instance.end(), instance.begin(), toupper);
+
+  return "METAL_" + name + "_" + instance;
+}
+
+uint64_t Device::base_address(const node &n) {
+  uint64_t b;
+
+  n.named_tuples(
+    "reg-names", "reg",
+    "control", tuple_t<target_addr, target_size>(), [&](target_addr base, target_size size) {
+	b = base;
+    });
+
+  return b;
+}
+
+void Device::emit_base(const node &n) {
+  os << "#define " << def_handle(n) << "_BASE_ADDRESS " << base_address(n) << "UL" << std::endl;
```

Comment:
It would be better to have these set of string labels define in say fdttag.h++
something like, #define BASE_ADDRESS_LABEL "BASE_ADDRESS"
"BASE_ADDRESS"
"SIZE"
"CLOCK_DIV"
"CLOCK_FREQUENCY"
etc

That way, metal_header/codes call also use. As is, metal_header and bare-header will out of sync, when they need to be in-sync in order to compile successfully

Reply:
I've implemented this change, what do you think now?